content/blog/2019-10-23-AutoVectoring/index.md

```diff
@@ -0,0 +1,242 @@
+    unsafe {
+        let a = _mm_load_si128(mem_a);
+        let b = _mm_load_si128(mem_b);
+        let c = _mm_add_epi32(a, b);
+        _mm_store_si128(mem_c, c);
+        std::mem::forget(mem_c)
+    }
+}
+```
+Consider the corresponding invocation of this Rust function in the interpreter for vector add.
+```gherkin=
+case "vadd": {
+    let vecA = getVec(instr, env, 0);
+    let vecB = getVec(instr, env, 1);
+    let vecC = new Int32Array(fixedVecSize);
+    vadd(vecA, vecB, vecC);
+    env.set(instr.dest, vecC);    
+    return NEXT;
+  }
+```
+These external calls add potentially significant overhead to the execution. We quantify this overhead to more accurately evaluate execution time in the interpreter. We run a vvadd benchmark, documented below, varying array sizes of (128, 1024, 2048, 4196, 8192). We run 10 iterations of each configuration and average the execution time between them. We then compare execution time with and without calls to the Rust library. We find there is a 16% overhead for making Rust calls. While we predict this will be offset for especially large arrays, we decided to add rust calls for serialized instructions as well to isolate the vectorized instructions as a variable in our experiment from the Rust call overhead variable.
+
+In addition, note the loads and stores that accompany the vectorized add instruction. The ```vadd``` function implements ```c[i] = a[i] + b[i]```. However, this line translated to Bril IR would look like the code below. 
+```
+va: vector = vload ai;
+vb: vector = vload bi;
+vc: vector = vadd va vb;
+vstore vc ci;
+```
+Ideally, these instructions would be in separate functions in our Rust library with separate invocations in the interpreter. However, TypeScript does not have a copmatible type for ```__m128i```, which represents a 128-bit SIMD register. In order to separate the ```vadd``` from ```vload``` and ```vstore```, we would either need to return a ```__m128i```, or invoke additional functions in the Rust SIMD crate to unpack this value into 4 32-bit integers. Without a compatible type for ```__m128i```, we did find a way to write a signature such that we could accept or return a ```__m128i``` from a single vector add. For the latter, adding functions to unpack integers adds potentially significant overhead every time we pass vectors between TypeScript and Rust. Therefore, we group these as a unit to more accurately mimic SIMD operations, as well as make vstore and vload effectively noops in the interpreter. This is valid as our implementation of automatic vectorization ensures each vload-vload-vstore is accompanied by some vector operation.
+
+
+## Evaluation
+### Correctness
+In order to verify correctness, we chose self-verifiying programs such that execution of the program can prove the program computed the same value, and thus is correct  in terms of inputs/outputs. These programs include vector-vector add, vector-vector multiply, and vector-vector subtract.
```

Comment:
That's fine, I suppose, but it would also have been interesting to try a program or two with a slightly different loop structure than just mapping a single scalar operand over all elements of two arrays.